pkg/login/social/generic_oauth.go

```diff
@@ -91,64 +92,97 @@ func (info *UserInfoJson) String() string {
 }
 
 func (s *SocialGenericOAuth) UserInfo(client *http.Client, token *oauth2.Token) (*BasicUserInfo, error) {
-	var data UserInfoJson
-	var err error
+	s.log.Debug("Getting user info")
+	tokenData := s.extractFromToken(token)
+	apiData := s.extractFromAPI(client)
 
 	userInfo := &BasicUserInfo{}
+	for _, data := range []*UserInfoJson{tokenData, apiData} {
+		if data == nil {
+			continue
+		}
 
-	if s.extractToken(&data, token) {
-		s.fillUserInfo(userInfo, &data)
-	}
+		s.log.Debug("Processing external user info", "source", data.source, "data", data)
+
+		if userInfo.Name == "" {
+			if data.Name != "" {
+				s.log.Debug("Setting user info name from name field")
+				userInfo.Name = data.Name
+			} else if data.DisplayName != "" {
+				s.log.Debug("Setting user info name from display name field")
+				userInfo.Name = data.DisplayName
+			}
+		}
+
+		if userInfo.Login == "" {
```

Comment:
Isn't better to extract this into separate function like `extractLogin(data *UserInfoJson)`?

Reply:
@alexanderzobnin I think it depends on personal preference; I follow the old rule of thumb of generally not making functions longer than the height of a screen, and until that point it's usually preferable to inline, because the more indirection you have the more difficult it's to review if the code is correct.

Reply:
It's probably not a big deal in this case because function is quite straightforward and it's easy to follow the code even if it's more than one screen.

Reply:
@alexanderzobnin Thanks, I take it you don't mind if I merge then? As you say, the code is relatively straightforward after all and not that difficult to follow.

Reply:
I'll merge it now - if you want, we can make another PR to refactor the function a little :)